### ISSUE TITLE
explore alternatives to global kernel lock

### ISSUE THREAD

**Issue by wjhun** [nanovms/nanos#1082]
With the introduction of #1078, the kernel uses a global lock to provide synchronization for kernel data and resources. While this allows simultaneous scheduling of unix threads across cores, it allows only one core to execute kernel code (syscalls and deferred operations off of the bhqueue and runqueue). Resources that require finer-grained locking (e.g. process memory map, physical page allocation, page table, driver ring populating) use spinlocks.

Explore the ground between these two extremes in an effort to increase parallelism of kernel code execution. For example, we could consider locks protecting certain classes of data, e.g. one for lwIP processing, one filesystem operations, etc.

**Comment by convolvatron** [comment]
i _think_ the biggest leverage here would be to allow independent sockets and the poll code to operate independently. unfortunately based my cursory look, it uses globals to manage the tcp input state while its being processed. so this would likely require a fork and partial refactor of the lwip code.

assume we did that, it might be possible to make this work without addressing the problem of thread-safe or concurrent (i.e per-cpu) allocation, as the work thats required on either the tcp input or tcp output path is fairly limited in scope. 

i guess i'm just assuming that filesystem i/o isn't a huge part of the workloads being targetted.

**Comment by convolvatron** [comment]
another potential structure here is to provide a per-system call map which says whether we're gated on acquiring the kernel lock for a particular system call. if/when a particular can be teased out from under the kernel lock, (something like futex or epoll, that may get its own concurrency control), we can turn this off.
